### ISSUE TITLE
Implement "Phase 1" of anonymous user upgrade

### ISSUE THREAD

**Issue by samtstern** [firebase/FirebaseUI-Android#1171]
This is related to #309 but because of the huge traffic / discussion over there I want to file a new clean issue about this piece.

For context this has already been release on Web and is coming in iOS:
https://github.com/firebase/firebaseui-web#upgrading-anonymous-users

This has two parts:

### 1 - New Flow Parameter

There will be a new boolean option to the intent builder, with the default as false:

```java
AuthUI.getInstance()
        .createSignInIntentBuilder()
        // ...
        .setAutoUpgradeAnonymousAccounts(true)
        .build();
```

Throughout the flow, wherever we previously called `signInWith...` or `signUpWith...` we will now do the following logic (pseudocode):

```java
if (hasAnonymousUser() && shouldAutoUpgrade) {
  linkAccount(...);
} else {
   signIn(...);
}
```

### 2 - Handling Link Failures

In the event that a link call fails, we will exit the flow with an error:

- New error code ANONYMOUS_UPGRADE_MERGE_CONFLICT
- The IdpResponse should contain the AuthCredential that we tried to link with but failed.
- The User in the IdpResponse will be the anonymous user.

When handling the error, the developer should:

- Read/cache the anonymous user data.
- Sign in with the AuthCredential.  They must make sure not to react to the auth state changed event.
- Write the anonymous user cached data as appropriate to the new user.
- Trigger their own sign-in success logic.

### Notes

In a future "phase" of this feature we can do something similar to what @SUPERCILEX has already done and handle the complete flow including data transfer.  In this "minimal" phase though this will only be for advanced developers that can handle the data transfer and display any relevant UI during that operation.

**Comment by atrepeklis** [comment]
Do you have a high level estimate on when this "Phase 1" will be released? I know we can always follow the SUPERCILEX approach but if we get it in a reasonable timeframe from the original source it might worth the wait.

**Comment by samtstern** [comment]
@atrepeklis sorry I don't have a good estimate right now.

**Comment by samtstern** [comment]
@SUPERCILEX I am going to take a crack at this, using #309 as my guide.

**Comment by SUPERCILEX** [comment]
@samtstern Awesome! I'm going to process my backlog tomorrow. 😊

**Comment by Phinatic** [comment]
@samtstern Do you have an active branch for this? Your approach above sounds simple and I was considering forking the repo to do this in my own app. I figured I would check for a branch here that may need a nudge to get it over the hump and approved instead of creating the fork.

**Comment by samtstern** [comment]
This has been released in version `4.1.0`

**Comment by johanbx** [comment]
Should this not be .`enableAnonymousUsersAutoUpgrade()`? I find no reference for `.setAutoUpgradeAnonymousAccounts(true)`
